### PR TITLE
Material Entry v1

### LINKED ISSUE #4949: [Material] Filled Entry Spec
### Description
https://material.io/design/components/text-fields.html#filled-text-field

image

### Expectations 

- **Visually Identical**
   The rendering should be such that the naked eye can’t tell the difference at a glance. _Not to be confused with “pixel perfect”_

- **States**
   Make sure these states are the same between ios/android and are true to the material guidelines (Inactive, active, focused, Activated, Hover, Disabled). Error messages will be added at a later date. Currently the color of underline and placeholder just uses the Material theme defaults. The material style use a predefined Alpha when the control is in a given state so this is currently what we hard code. We will expose this through an attached property at a later date. The renderers are constructed in such a way that the user can override the styling if they want to customize these color states

### Stretch goal
- The ability to add a trailing/leading icon. 

# Scenarios
```XAML
<Entry Visual="Material" />
```

```C#
new Editor() { Visual = VisualMarker.Material }
```

### Current issues
- iOS has bottom padding below control that Android doesn't have
- disabled coloring slightly different between each
- control itself renderers with different height on iOS vs Android
- Font/Font sizes are slightly different between two controls
- underline colors are subtly different

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###
Styles changes for ios and android to make Material Entry 
  

### Known Issues
- The height on android renders differently based on if it starts with text vs not

### Issues Resolved ### 
- fixes #4949 

### Platforms Affected ### 

- Core/XAML (all platforms)
- iOS
- Android

### Before/After Screenshots ### 

image

### Testing Procedure ###
Play around in the gallery:
- Entry Gallery (Material)
- Entry Material Demos

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by PureWeen** [comment]
build --uitests

File: Xamarin.Forms.Platform.iOS/Renderers/EntryRenderer.cs [hunk 1]
```diff
@@ -116,11 +116,12 @@ protected override void OnElementChanged(ElementChangedEventArgs<Entry> e)
 			_cursorPositionChangePending = Element.IsSet(Entry.CursorPositionProperty);
 			_selectionLengthChangePending = Element.IsSet(Entry.SelectionLengthProperty);
 
+			// Font needs to be set before Text and Placeholder so that they layout correctly when set
```
Comment (PureWeen) [inline Xamarin.Forms.Platform.iOS/Renderers/EntryRenderer.cs]: this is the only change that could in theory cause breaking change to existing behavior

File: Xamarin.Forms.Platform.iOS/Xamarin.Forms.Platform.iOS.csproj [hunk 2]
```diff
@@ -113,6 +113,9 @@
     <Compile Include="$(MSBuildThisFileDirectory)Renderers\WebViewRenderer.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Properties\AssemblyInfo.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Extensions\CellExtensions.cs" />
+    <Compile Include="..\Xamarin.Forms.Platform.Android\Material\MaterialColors.cs">
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply Xamarin.Forms.Platform.iOS/Xamarin.Forms.Platform.iOS.csproj]: I went back and forth on the best way to implement this one. I'm not super excited about just using the same file like this but for now I'd like to delay adding this into a cross platform place until I can think a bit more on the best way to expose. Once we start adding material based attached properties or themes then we'll move it somewhere better. For now this let's me have the code used by both platforms in the same place but delay exposing any APIs out

**Comment by PureWeen** [comment]
build --uitests

**Comment by PureWeen** [comment]
@paymicro  can you open up a new issue with the two things you found and assign them to me please?

**Comment by mattleibow** [comment]
Could the tab thing have to do with https://github.com/xamarin/Xamarin.Forms/pull/4703 or https://github.com/xamarin/Xamarin.Forms/pull/4612?
Those were recently merged, so maybe we need to test again.

**Comment by paymicro** [comment]
> @paymicro can you open up a new issue with the two things you found and assign them to me please?

@PureWeen  sure. I don't have right to assign. Here are the links: https://github.com/xamarin/Xamarin.Forms/issues/5102 https://github.com/xamarin/Xamarin.Forms/issues/5103

File: Xamarin.Forms.Platform.Android/Material/MaterialColors.cs [hunk 3]
```diff
@@ -1,12 +1,88 @@
-﻿using Android.Content.Res;
+﻿// once we implement material configurations on core elements this can all be moved up to that
+// for now just leaving this as an internal class to make matching colors between two platforms easier
+
+#if __ANDROID__
+using Android.Content.Res;
 using AColor = Android.Graphics.Color;
+#else
+using MaterialComponents;
+using AColor = UIKit.UIColor;
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply Xamarin.Forms.Platform.Android/Material/MaterialColors.cs]: Agreed! I meant to make this change before the merge but was waiting for the UI Tests to finish before pushing. I'll update the naming to something better.

**Pull request merged**